[FLINK-8375][network] Remove unnecessary synchronization

pnowojski opened the pull request:
Synchronized blocks in ResultPartition could affect only:
1. totalNumberOfBuffers and totalNumberOfBytes counters
2. subpartition add(), finish() and release() calls.

However:
1. accessors to the counters where not synchronized.
2a. add(), finish() and release() methods for PipelinedSubpartition were already threads safe
2b. add(), finish() and release() methods for SpillableSubpartition were made thread safe in
this commit, by basically pushing synchronized section down one level.

NicoK commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java:
```diff
@@ -275,13 +275,11 @@ public void writeBuffer(Buffer buffer, int subpartitionIndex) throws IOException
-			}
+			// Update statistics
+			totalNumberOfBuffers++;
+			totalNumberOfBytes += buffer.getSize();
```
Please note that without the output flusher thread being removed, we do have two threads increasing these integers which may produce garbage. (btw: should we switch `totalNumberOfBuffers` to a `long`?)

NicoK reviewed (commented):
Please note that the thread-safety of `PipelinedSubpartition` wasn't as simple as relying on `synchronized(buffers)` in its methods. `add()`, `finish`, and `release()` do acquire the read view under this lock but then work with it afterwards which could come in any arbitrary order, e.g. the notification of a new buffer inside `add()` could actually refer to a released channel (with all buffers recycled) once it arrives but
- `PartitionRequestQueue#writeAndFlushNextMessageIfPossible` handles this for obvious reasons (the notification only happens when switching from empty to non-empty, but `release()` may be called any time).
- `LocalInputChannel` works similarly in that it's `getNextBuffer()` method throws a `CancelTaskException` when a polled buffer is `null` and the channel was released. I guess, that is intended to shut down the receiver as well.

NicoK reviewed (changes requested):
Regarding `SpillableSubpartition`, the PR does not change the current behaviour and we can make the optimisation there a separate task. So this is only about removing unnecessary synchronization for the `PipelinedSubpartition` code path used in streaming jobs.

Since the (potential) implications of this change are quite big, it should get an own JIRA ticket for it (and so should the separate task for optimising `SpillableSubpartition`).

The statistics in `ResultPartition` should be fixed though, at the simplest by removing them (they are currently unused and we can retrieve them from enumerating the subpartitions' statistics (or the `RecordWriter`) if we decide to use them).

pnowojski committed:
[FLINK-8375][network] Remove unnecessary synchronization

Synchronized blocks in ResultPartition could affect only:
1. totalNumberOfBuffers and totalNumberOfBytes counters
2. subpartition add(), finish() and release() calls.

However:
1. counters were not used anywhere - they are removed by this commit
2a. add(), finish() and release() methods for PipelinedSubpartition were already threads safe
2b. add(), finish() and release() methods for SpillableSubpartition were made thread safe in
this commit, by basically pushing synchronized section down one level.

pnowojski commented:
Done, created https://issues.apache.org/jira/browse/FLINK-8375 and https://issues.apache.org/jira/browse/FLINK-8376

Please take a look and let me know if it has `LGTM`

NicoK commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/ResultPartitionTest.java:
```diff
@@ -192,8 +192,6 @@ public void testWriteBufferToAllSubpartitionsReferenceCounting() throws Exceptio
 		partition.writeBufferToAllSubpartitions(buffer);
 
-		// Verify added to all queues, i.e. two buffers in total
-		assertEquals(2, partition.getTotalNumberOfBuffers());
```
How about checking this with `ResultPartition#getAllPartitions()` for each subpartition? Otherwise this test doesn't really test anything.

pnowojski replied:
Actually it is testing exactly what the test's name suggests. It sets up partition with two subpartitions and asserts `assertTrue(buffer.isRecycled());`  after `writeBufferToAllSubpartitions(buffer)`. 

`assertEquals(2, partition.getTotalNumberOfBuffers());` 

was:

1. first of all redundant (hence it would be copied/paste code), since the fact whether buffer was written to both subpartitions or not is surely tested somewhere else in some broadcast emit unit test.
2. not checking exactly what you intended, since `writeBufferToAllSubpartitions(...)` implemented in one of many buggy ways would still pass this test. To actually test what this check was intended to do, we should verify that the `buffer` passed to `writeBufferToAllSubpartitions(buffer)` was actually *written* to both partitions - not that statistics were accounted for it, but again, see point 1.

NicoK reviewed (changes requested):
thanks for the changes, `ResultPartitionTest#testWriteBufferToAllSubpartitionsReferenceCounting()` however needs to be adapted further
